components/ai_chat/core/browser/model_service_unittest.cc

```diff
@@ -9,36 +9,167 @@
+      observer_->Observe(service_.get());
+    }
+    return service_.get();
   }
 
-  void TearDown() override { OSCryptMocker::TearDown(); }
+  void TearDown() override {
+    OSCryptMocker::TearDown();
+    observer_.reset();
+  }
 
+ protected:
   TestingPrefServiceSimple pref_service_;
+  std::unique_ptr<NiceMock<MockModelServiceObserver>> observer_;
+
+ private:
   std::unique_ptr<ModelService> service_;
 };
 
+class ModelServiceTestWithDifferentPremiumModel : public ModelServiceTest {
+ public:
+  ModelServiceTestWithDifferentPremiumModel() {
+    scoped_feature_list_.InitAndEnableFeatureWithParameters(
+        features::kAIChat,
+        {
+            {features::kAIModelsDefaultKey.name, "chat-leo-expanded"},
+            {features::kAIModelsPremiumDefaultKey.name, "claude-3-sonnet"},
+        });
+  }
+
+ private:
+  base::test::ScopedFeatureList scoped_feature_list_;
+};
+
+class ModelServiceTestWithSamePremiumModel : public ModelServiceTest {
+ public:
+  void SetUp() override {
+    scoped_feature_list_.InitAndEnableFeatureWithParameters(
+        features::kAIChat,
+        {
+            {features::kAIModelsDefaultKey.name, "chat-leo-expanded"},
+            {features::kAIModelsPremiumDefaultKey.name, "chat-leo-expanded"},
+        });
+    ModelServiceTest::SetUp();
+  }
+
+  base::test::ScopedFeatureList scoped_feature_list_;
+};
+
+TEST_F(ModelServiceTest, MigrateOldClaudeDefaultModelKey) {
+  // Set default to the old key for claude
+  pref_service_.SetString("brave.ai_chat.default_model_key",
+                          "chat-claude-instant");
+  // Call Migrate even though it shouldn't touch this pref value, precisely
+  // to test that it doesn't interfere with the translation from old claude to
+  // new claude.
+  ModelService::MigrateProfilePrefs(&pref_service_);
+  // Verify uses non-premium version
+  EXPECT_EQ(GetService()->GetDefaultModelKey(), "chat-claude-instant");
+  // Verify uses premium version
+  GetService()->OnPremiumStatus(mojom::PremiumStatus::Active);
+  EXPECT_EQ(GetService()->GetDefaultModelKey(), "chat-claude-sonnet");
+}
+
+TEST_F(ModelServiceTest, MigrateOldClaudeDefaultModelKey_OnlyOnce) {
+  // Set default to the old key for claude
+  pref_service_.SetString("brave.ai_chat.default_model_key",
+                          "chat-claude-instant");
+  // Call Migrate even though it shouldn't touch this pref value, precisely
+  // to test that it doesn't interfere with the translation from old claude to
+  // new claude.
+  ModelService::MigrateProfilePrefs(&pref_service_);
+  // Verify uses non-premium version
+  EXPECT_EQ(GetService()->GetDefaultModelKey(), "chat-claude-instant");
+  // Verify keeps non-premium version
+  GetService()->OnPremiumStatus(mojom::PremiumStatus::Inactive);
+  EXPECT_EQ(GetService()->GetDefaultModelKey(), "chat-claude-instant");
+  GetService()->OnPremiumStatus(mojom::PremiumStatus::Active);
+  EXPECT_EQ(GetService()->GetDefaultModelKey(), "chat-claude-instant");
```

Comment:
shouldn't we also check `OnDefaultModelChanged`? same with test suite `MigrateOldClaudeDefaultModelKey`, `MigrateToPremiumDefaultModel` and `MigrateToPremiumDefaultModel_None` whichever calls `OnPremiumStatus`

Reply:
Thanks - added the event checks, and actually found a bug in the test and the tested code - models weren't being loaded before migration from `chat-claude-instant` pref value, so the intended replacements (chat-claude-haiku) weren't being found and the migration was ignored. The test case had the wrong value so was passing.